feat: `--all-devices` and `--device-type` cli flags

joegoldin opened the pull request:
Compiled on linux (via WSL) and Windows and tested on Windows with a Litra Beam LX and a Litra Glow. All commands worked as expected.

<details>
<summary>CLI Testing Output</summary>

```
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): Off 🌑
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (009F6369D08D): Off 🌑
  - Brightness: 56 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe toggle --all-devices        
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): On 💡
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (009F6369D08D): On 💡
  - Brightness: 56 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe toggle -s 009F6369D08D      
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): On 💡
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (009F6369D08D): Off 🌑
  - Brightness: 56 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe toggle -s 2229FE006UH8      
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): Off 🌑
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (009F6369D08D): Off 🌑
  - Brightness: 56 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe toggle --device-type LitraBeamLX
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): Off 🌑
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (009F6369D08D): On 💡
  - Brightness: 56 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe toggle --device-type LitraGlow
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): On 💡
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (009F6369D08D): On 💡
  - Brightness: 56 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe brightness --device-type LitraBeamLX --percentage 50
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): Off 🌑
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (009F6369D08D): Off 🌑
  - Brightness: 215 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe brightness --device-type LitraBeamLX --percentage 20
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): Off 🌑
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (009F6369D08D): Off 🌑
  - Brightness: 104 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe brightness -s 009F6369D08D --percentage 100
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): Off 🌑
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (009F6369D08D): Off 🌑
  - Brightness: 400 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe temperature -s 009F6369D08D --value 5000
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): Off 🌑
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (009F6369D08D): Off 🌑
  - Brightness: 400 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe temperature -s 009F6369D08D --value 4300
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): Off 🌑
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (009F6369D08D): Off 🌑
  - Brightness: 400 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
```
</details>

timrogers commented on src/lib.rs:
```diff
@@ -217,9 +219,37 @@ impl DeviceHandle {
+    /// serial is generated based on device type, product ID, and path.
     pub fn serial_number(&self) -> DeviceResult<Option<String>> {
         match self.hid_device.get_device_info() {
-            Ok(device_info) => Ok(device_info.serial_number().map(String::from)),
```
How would feel about skipping this part, and instead separately exposing a "device path" and a way of targeting a device by its path rather than serial number?

joegoldin replied:
That works for me!

timrogers commented on src/main.rs:
```diff
@@ -20,22 +20,38 @@ enum Commands {
     On {
         #[clap(long, short, help = "The serial number of the Logitech Litra device")]
         serial_number: Option<String>,
+        #[clap(long, help = "Select devices by their type (LitraGlow, LitraBeam, LitraBeamLX)")]
```
I think we should add a validation when running all the commands to ensure that exactly zero or one of these filters is applied, and not multiple. I think that will be easier to reason about.

timrogers replied:
Is 
```suggestion
        #[clap(long, help = "Specify the type of device to target. All connected devices of the provided type will be targeted. ")]
```

timrogers commented on src/main.rs:
```diff
@@ -20,22 +20,38 @@ enum Commands {
         #[clap(long, short, help = "The serial number of the Logitech Litra device")]
         serial_number: Option<String>,
+        #[clap(long, help = "Select devices by their type (LitraGlow, LitraBeam, LitraBeamLX)")]
+        device_type: Option<String>,
```
Is there a way we can use the `DeviceType` enum here rather than `String`?

timrogers commented on src/main.rs:
```diff
@@ -153,14 +189,33 @@ fn get_is_on_emoji(is_on: bool) -> &'static str {
-                .serial_number()
-                .is_some_and(|actual| &actual == expected)
-        })
+        // Check device type if specified
```
I guess we can simplify this method if we only allow one filter at a time.

timrogers commented on src/main.rs:
```diff
@@ -20,22 +20,38 @@ enum Commands {
     On {
         #[clap(long, short, help = "The serial number of the Logitech Litra device")]
```
`Specify the device to target by its serial number`

timrogers commented on src/main.rs:
```diff
@@ -20,22 +20,38 @@ enum Commands {
+        #[clap(long, help = "Select devices by their type (LitraGlow, LitraBeam, LitraBeamLX)")]
+        device_type: Option<String>,
+        #[clap(long, help = "Apply command to all connected devices", default_value = "false")]
+        all_devices: bool,
```
Rather than do this, I wonder if we should change the default so we target all devices by default, and only limit the selection of a serial number or device type is provided? 

What do you think? That would be a breaking change, but I think it's a reasonable one.

joegoldin replied:
I had originally considered that but opted for backwards compat -- I think it's a totally reasonable change, so I did that in the latest commit :)

timrogers reviewed (changes requested):
Loving the direction on this so far! I have a few suggestions.

joegoldin commented:
- Removed deterministic serial number generation and implement device path targeting
- Added validation to ensure only one filter can be applied at a time
- Used DeviceType enum instead of String for device_type parameter
- Changed default behavior to target all devices
- Updated help text copy
- Updated README

<details>
<summary>CLI Testing Output</summary>

```
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices                                                   
- Litra Glow (2229FE006UH8): Off 🌑
  - Device path: \\?\HID#VID_046D&PID_C900&Col02#7&2d3520f0&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (UNKNOWN): Off 🌑
  - Device path: \\?\HID#VID_046D&PID_C903&Col02#7&2a2a87de&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
  - Brightness: 67 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe toggle -p "\\?\HID#VID_046D&PID_C900&Col02#7&2d3520f0&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}"
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices                                                   
- Litra Glow (2229FE006UH8): On 💡
  - Device path: \\?\HID#VID_046D&PID_C900&Col02#7&2d3520f0&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (UNKNOWN): Off 🌑
  - Device path: \\?\HID#VID_046D&PID_C903&Col02#7&2a2a87de&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
  - Brightness: 67 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe toggle -p "\\?\HID#VID_046D&PID_C903&Col02#7&2a2a87de&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}"
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices                                                   
- Litra Glow (2229FE006UH8): On 💡
  - Device path: \\?\HID#VID_046D&PID_C900&Col02#7&2d3520f0&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (UNKNOWN): On 💡
  - Device path: \\?\HID#VID_046D&PID_C903&Col02#7&2a2a87de&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
  - Brightness: 67 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe toggle
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): Off 🌑
  - Device path: \\?\HID#VID_046D&PID_C900&Col02#7&2d3520f0&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (UNKNOWN): Off 🌑
  - Device path: \\?\HID#VID_046D&PID_C903&Col02#7&2a2a87de&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
  - Brightness: 67 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe toggle -t glow                                            
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices       
- Litra Glow (2229FE006UH8): On 💡
  - Device path: \\?\HID#VID_046D&PID_C900&Col02#7&2d3520f0&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (UNKNOWN): Off 🌑
  - Device path: \\?\HID#VID_046D&PID_C903&Col02#7&2a2a87de&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
  - Brightness: 67 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe toggle -t beamlx
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices
- Litra Glow (2229FE006UH8): On 💡
  - Device path: \\?\HID#VID_046D&PID_C900&Col02#7&2d3520f0&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
  - Brightness: 20 lm
    - Minimum: 20 lm
    - Maximum: 250 lm
  - Temperature: 5000 K
    - Minimum: 2700 K
    - Maximum: 6500 K
- Litra Beam LX (UNKNOWN): On 💡
  - Device path: \\?\HID#VID_046D&PID_C903&Col02#7&2a2a87de&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
  - Brightness: 67 lm
    - Minimum: 30 lm
    - Maximum: 400 lm
  - Temperature: 4300 K
    - Minimum: 2700 K
    - Maximum: 6500 K
PS C:\Users\Joe\Development\litra-rs> .\target\debug\litra.exe devices --json
[{"serial_number":"2229FE006UH8","device_path":"\\\\?\\HID#VID_046D&PID_C900&Col02#7&2d3520f0&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}","device_type":"Litra Glow","is_on":true,"brightness_in_lumen":20,"temperature_in_kelvin":5000,"minimum_brightness_in_lumen":20,"maximum_brightness_in_lumen":250,"minimum_temperature_in_kelvin":2700,"maximum_temperature_in_kelvin":6500},{"serial_number":"UNKNOWN","device_path":"\\\\?\\HID#VID_046D&PID_C903&Col02#7&2a2a87de&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}","device_type":"Litra Beam LX","is_on":true,"brightness_in_lumen":67,"temperature_in_kelvin":4300,"minimum_brightness_in_lumen":30,"maximum_brightness_in_lumen":400,"minimum_temperature_in_kelvin":2700,"maximum_temperature_in_kelvin":6500}]
```

</details>

joegoldin commented:
Hey @timrogers, just a friendly bump here :)

timrogers commented:
Thanks for the reminder on this! I've been very busy recently but should get to it in the next couple of days.
